drivers/adc/ad463x/iio_ad463x.c

```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+ * @param buf	  - Command buffer to be filled with requested data.
+ * @param len     - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv    - Command attribute id.
+ * @return ret    - Result of the reading procedure.
+ *		    In case of success, the size of the read data is returned.
+*/
+static int ad463x_iio_store_scale(void *dev, char *buf, uint32_t len,
+				  const struct iio_ch_info *channel, intptr_t priv)
+{
+	struct iio_ad463x *iio_desc = dev;
+	struct ad463x_dev *ad463x_dev;
+	ad463x_dev = iio_desc->ad463x_desc;
+
+	int32_t vals[2];
+	/** Scale only configurable if PGIA is available */
+	if(ad463x_dev->has_pgia) {
+		iio_parse_value(buf, IIO_VAL_FRACTIONAL, &vals[0], &vals[1]);
+
+		int32_t gain_idx = ad463x_calc_pgia_gain(vals[0], vals[1], ad463x_dev->vref,
+				   ad463x_dev->real_bits_precision);
+		return ad463x_set_pgia_gain(ad463x_dev, (uint8_t)gain_idx);
+	}
+	return 0;
+}
+
+/**
+ * @brief Handles the read request for scale available attribute.
+ * @param dev     - The iio device structure.
+ * @param buf	  - Command buffer to be filled with requested data.
+ * @param len     - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv    - Command attribute id.
+ * @return ret    - Result of the reading procedure.
+ *		    In case of success, the size of the read data is returned.
+*/
+static int ad463x_iio_read_scale_avail(void *dev, char *buf,
+				       uint32_t len, const struct iio_ch_info *channel, intptr_t priv)
+{
+	int32_t vals[2];
+	char buffer[20];
+	struct iio_ad463x *iio_desc = dev;
+	struct ad463x_dev *ad463x_dev;
+	ad463x_dev = iio_desc->ad463x_desc;
+
+	if(ad463x_dev->has_pgia) {
+		/** Remove data existing in the buffer */
+		strcpy(buf, "");
+
+		/** Go through the values in ODR table and add them to the buffer */
+		for (uint8_t i = 0; i < NO_OS_ARRAY_SIZE(ad463x_dev->scale_table); i++) {
+			vals[0] = ad463x_dev->scale_table[i][0];
+			vals[1] = ad463x_dev->scale_table[i][1];
+			iio_format_value(buffer, sizeof(buffer), IIO_VAL_INT_PLUS_NANO, 2, vals);
+			strcat(buf, buffer);
+			strcat(buf, " ");
+		}
+		return strlen(buf);
+	}
+
+	vals[0] = (ad463x_dev->vref * 2) / 1000;
+	vals[1] = ad463x_dev->capture_data_width;
+	return iio_format_value(buf, len, IIO_VAL_FRACTIONAL_LOG2, 2, vals);
+}
+
+/**
+ * @brief Reads the number of given samples for the selected channels
+ * @param dev     - The iio device structure.
+ * @param buf	  - Command buffer to be filled with requested data.
+ * @param samples - Number of samples to be returned
+ * @return ret    - Result of the reading procedure.
+ * 					In case of success, the size of the read data is returned.
+*/
 static int32_t _iio_ad463x_read_dev(struct iio_ad463x *desc, uint32_t *buff,
 				    uint32_t nb_samples)
 {
 	int ret;
-	uint32_t data[nb_samples], i, j, ch;
+	uint16_t total_samples = nb_samples * 2;
+	uint32_t data[total_samples], i, j, ch;
```

Comment:
It's not usually that great to use the stack for potentially large arrays like that. If you really need the `data` array, allocate the memory on the heap using `no_os_calloc()`.